mutation: Ignore dummy rows when consuming clustering fragments

mikolajsieluzycki opened the pull request:
consume_clustering_fragments already ignores dummy rows, but does it in
the wrong place. Currently they're ignored after comparing them with
range tombstones. This change skips them before any useful work is done
with them.

Consider a simplified mutation reversal scenario scenario (ckp is
clustering key prefix, -1, 0, 1 are bound_weights):

schema_ptr s = schema_builder{"ks", "cf"}
    .with_column("pk", bytes_type, column_kind::partition_key)
    .with_column("ck1", bytes_type, column_kind::clustering_key)
    .build();

Input range tombstone positions:
    {clustered, ckp{}, before}
    {clustered, ckp{1}, after}

Clustering rows:
    {clustered, ckp{2}, equal}
    {clustered, ckp{}, after} // dummy row

During reversal, clustering rows are read backwards, and reversed range
tombstone positions are read forwards (because the range tombstones are
reversed and applied backwards). The read order in the example above is:

Reversed range tombstone positions:
    1: {clustered, ckp{}, before}
    2: {clustered, ckp{1}, before}

Clustering rows read backwards:
    3: {clustered, ckp{}, after} // dummy row
    4: {clustered, ckp{2}, equal}

Then we effectively do the merge part of merge sort, trying to put all
fragments in order according to their positions from the two lists
above. However, the dummy row is used in the comparison, and it compares
to be gt each of the reversed range tombstone positions. Then we
try to emit the clustering row, but only at that point we notice it's
dummy and should be skipped. Subsequent row with ckp{2} is compared to
the last used range tombstone position and the fragments are out of
order (in reversed schema, ckp{2} should come before ckp{1}).

The solution is to move the logic skipping the dummy clustering rows to
the beginning of the loop, so they can be ignored before they're used.

Fixes: https://github.com/scylladb/scylla/issues/11147

bhalevy commented:
For
```
Input range tombstone positions:
{clustered, ckp{}, before}
{clustered, ckp{1}, after}
```

Shouldn't this be in reverse (2. before 1.)?
```
Reversed range tombstone positions:
    1: {clustered, ckp{}, before}
    2: {clustered, ckp{1}, before}
```

bhalevy commented:
Otherwise looks good to me

avikivity commented:
Looks reasonable, but from the description it appears that the bug does have a real effect on results. If that's correct, then we should have a targeted test and an issue #.

If not, then I'd like to understand why.

mikolajsieluzycki commented:
The change was inspired by trying to enforce the assumption inside `range_tombstone_change_generator::flush` that `_lower_bound < upper_bound`. I have a targeted test which trips this assertion and passes with the change proposed. That being said, this assertion is tripped in other situations, so I have to fix them first. I'll try to modify the test so that it fails without any additional changes added and create an issue.

mikolajsieluzycki commented:
@bhalevy My description is correct, but incomplete. Positions are taken from range tombstones. Individual range tombstones are reverted (end is swapped with start), and they're applied in inverse order. In the description I only provided starts of range tombstones, I should update it with full description.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1514/

denesb commented:
One detail missing from the cover letter is that the position of rows is not reversed when reading in reverse. Reversing a position would transform an `after(ckp{})` to `before({ckp})`. We do this for range tombstones but **not** for rows, the reason for this is that rows always have `equal(ckp{})` positions, that are identical to their reverse counterparts. Dummy rows violate this convention and this is what has bitten us.
Another possible fix is to just reverse the position of rows too, automatically fixing the mis-placement of dummy rows. But ignoring them early is probably the better fix.

denesb reviewed (commented):
Fix itself LGTM, but we need a test, like @avikivity requested.

mikolajsieluzycki commented:
In v2:
- Changed the commit message to better describe the range tombstones and note that we don't reverse the positions of dummy clustering rows
- Added two tests - one handcrafted and one random that ensure that mutations are consumed in order (which was the underlying problem here)

denesb commented on test/boost/mutation_test.cc:
```diff
@@ -2817,3 +2817,104 @@ SEASTAR_THREAD_TEST_CASE(test_compactor_range_tombstone_spanning_many_pages) {
+    });
+}
+
+SEASTAR_TEST_CASE(mutation_is_consumed_in_monotonic_order) {
```
Isn't this the same as `test_mutation_consume_position_monotonicity`?

mikolajsieluzycki replied:
Indeed looks very similar. I'll drop the duplicate, pull out the validator from that test and reuse it in the handcrafted one

mikolajsieluzycki commented:
v3:
- Extracted validating consumer from `test_mutation_consume_position_monotonicity` and use it in the handcrafted test
- Remove the fuzz test, as it's a duplicate of `test_mutation_consume_position_monotonicity`

bhalevy commented on test/boost/mutation_test.cc:
```diff
@@ -2531,6 +2531,37 @@ SEASTAR_THREAD_TEST_CASE(test_mutation_consume) {
+        BOOST_REQUIRE(_validator.on_end_of_stream());
+    }
+};
+
```
it seems appropriate to move this to test/lib/ maybe to a new header file e.g. mutation_fragment_stream_validator.hh
so it can be used also elsewhere down the road.

bhalevy replied:
or to mutation_assertions.hh that already calls BOOST_* stuff.

bhalevy commented on test/boost/mutation_test.cc:
```diff
@@ -2597,6 +2597,50 @@ SEASTAR_THREAD_TEST_CASE(test_mutation_consume_position_monotonicity) {
+        m.apply(mutation_fragment(*s, p, std::move(cr3)));
+        m.partition().ensure_last_dummy(*s);
+        mutation m1{m};
+        mutation m2{m};
```
m2 seems to be unused.

mikolajsieluzycki replied:
Good catch, I only need 2 mutations - one for normal and one for reverse check

mikolajsieluzycki commented:
v4:
- Moved `validating_consumer` to `test/lib/mutation_assertions.hh`
- Removed unnecessary `m2` from the test

bhalevy reviewed (approved):
LGTM

denesb commented:
Will queue once CI approves.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1534/

denesb commented:
We need an issue for this in case we decide to backport it. Please open one if there isn't one already, and add the appropriate 
"Fixes" to the cover letter.

denesb commented:
(I queued and de-queued the series because of course I remembered to check the issue right after pushing)

tgrabiec commented on mutation.hh:
```diff
@@ -267,6 +267,12 @@ std::optional<stop_iteration> consume_clustering_fragments(schema_ptr s, mutatio
+        }
         bool emit_rt;
         if (crs_it != crs_end && rts_it != rts_end) {
             const auto cmp_res = cmp(rts_it->position(), crs_it->position());
```
A more generic solution would be to use `crs_it->position().reversed()` here for reverse schema.

Your solution works too, but it assumes that all !dummy() positions are invariant to the clustering order transformation (can be used in both forward and reverse schema). Currently holds, but again, relying on fewer implicit assumptions is good for code stability.
